src/core/service.c

```diff
@@ -3065,7 +3139,10 @@ static int service_reload(Unit *u) {
 
         assert(IN_SET(s->state, SERVICE_RUNNING, SERVICE_EXITED));
 
-        service_enter_reload(s);
+        if (s->type == SERVICE_NOTIFY_RELOAD)
+                service_enter_reload_by_notify(s);
+        else
+                service_enter_reload_by_exec(s);
```

Comment:
this is backwards. you are deciding on the 2nd step here before doing the 1st, no?

also this changes behaviour quite drastically if you combine ExecReload= with Type=notify-reload, because previously we'd send the signal *and* execute the specified commands, and not you just ignore the commands.

Reply:
But that was never supported? We explicitly document it as an alternative, not something to use together: "Using this service type and implementing this reload protocol is an efficient alternative to providing an ExecReload= command for reloading of the service's configuration."

https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html

I did a check for everything on Github using notify-reload, and nothing also uses the command at the same time

Reply:
It is supported and you can combine the two right now. Please retain the compat.

Reply:
But the documentation says they are alternative, all the users that can be found use it as alternative, the PR that introduced it said it was an alternative:

> Well, service authors can drop ExecReload= if they use Type=notify-reload.

https://github.com/systemd/systemd/pull/25916#issuecomment-1383782081

So to me this looks like a bug fix. But it is unrelated and just overlapped, so I'll open a separate PR for this after this is merged and undo it here

Reply:
I am pretty sure we should allow the combo of ExecReload= and Reload=notify-reload. People might have valid reasons why they also want to run some code after reload, even if the primary way to reload is smart.

The way i see it, we'll eventually have to have a four stage reload: 

1. reloading confexts
2. refreshing creds
3. notify/reload signal stuff
4. ExecReload=

(and probably in this order)

i do not see any reason we should only allow a subset of these combinations.

Reply:
I think it's a really bad idea to mix those. But that's for another PR, I've restored the original behaviour in this one.